DBZ-5847: Add ComputePartition Transformation. (#4089)
debezium/debezium: documentation/modules/ROOT/pages/transformations/partition-routing.adoc

:page-aliases: configuration/partition-routing.adoc
// Category: debezium-using
// Type: assembly
// ModuleID: route-record-to-specific-partition-base-on-specific-data-collections-column
// Title: Partition Routing
[id="partition-routing"]
= Partition Routing

:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highlighter: highlight.js

toc::[]

As described in xref:install.adoc#configuring-debezium-topics[Installing Debezium], by default, data collections changes will be sent to a topic with a single partition.
If you decide to change this configuration as described in xref:configuration/topic-auto-create-config.adoc[Customization of Kafka Connect automatic topic creation], Kafka will determine the partition by hashing the key by default.

> You can use this SMT if you want to use a specific data collection column to calculate the destination partition. The hash of the value will be used to compute the destination partition.

Reviewer: ```suggestion
However, in some cases, you might want {prodname} to  route events to a specific partition. 
The partition routing SMT enables you to route events to specific destination partitions based on a specified column name value. {prodname} uses the hash of the specified value to determine the destination partition.
```

// Type: concept

> // Title: Example: {prodname} basic partition routing SMT configuration
> // ModuleID: example-basic-debezium-partition-routing-smt-configuration

Reviewer: ```suggestion
// Title: Example: Basic configuration of the {prodname} partition routing SMT
// ModuleID: basic-configuration-of-the-debezium-partition-routing-smt
```

[[example-basic-partition-routing-configuration-example]]
== Example: Basic configuration

You configure the partition routing transformation in the {prodname} connector's Kafka Connect configuration.

> In the configuration, you specify the data collection column you want to use to calculate the partition and the max number of partition configured for the data collection.

Reviewer: ```suggestion
The configuration specifies the following parameters:

* The data collection column to use to calculate the destination partition.
* The maximum number of partitions permitted for the data collection.
```

> Only events that come from configured data collections will be processed. The others will be simply leaved as is.

Reviewer: ```suggestion
The partition routing SMT processes only events that originate from the specified data collections.
Events from other data collections are ignored and are routed to their default destinations.
```

To configure a {prodname} connector to route events to a specific partition, configure the `ComputePartition` SMT in the Kafka Connect configuration for the {prodname} connector.

For example, you might add the following configuration in your connector configuration.

[source]
----
...
topic.creation.default.partitions=2
topic.creation.default.replication.factor=1
...
transforms=ComputePartition
transforms.ComputePartition.type=io.debezium.transforms.partitions.ComputePartition
transforms.ComputePartition.partition.data-collections.field.mappings=inventory.products:name,inventory.orders:purchaser
transforms.ComputePartition.partition.data-collections.partition.num.mappings=inventory.products:2,inventory.orders:2
...
----

> The preceding example specifies to enable partition routing computation for `products` and `orders` tables.
> It specifies that `name` column will be used to compute the partition for the `products` data-collections and that the number of partition for this data-collections is `2`.
> Note that the number of partitions must be the number that you have configured for the topic. As you can see in the example we have defined that every topic will have `2` partitions.

Reviewer: ```suggestion
The configuration in the preceding example applies to events that originate from the `products` and `orders` {data-collection}s.
It specifies that `name` column will be used to compute the partition for the `products` data-collections and that the number of partition for this data-collections is `2`.
The value that you specify for the number of partitions must match the number of partitions are configured for the topic. 
For example, in the example, `topic.creation.default.partitions` creates topics  that have `2` partitions.
```

> Given this `Products` table

Reviewer: ```suggestion
To illustrate the behavior of the SMT, we can look at how it processes data from the following sample {data-collection}.
```

> .Products table

Reviewer: ```suggestion
.Products {data-collection}
```

[cols="25%a,25%a,25%a,25%a"]
|===
|id
|name
|description
|weight

|101
|scooter
|Small 2-wheel scooter
|   3.14

|102
|car battery
|12V car battery
|   8.1
|103
|12-pack drill bits
|12-pack of drill bits with sizes ranging from #40 to #3
|   0.8
|104
|hammer
|12oz carpenter's hammer
|  0.75
|105
|hammer
|14oz carpenter's hammer
| 0.875
|106
|hammer
|16oz carpenter's hammer
|   1.0
|107
|rocks
|box of assorted rocks
|   5.3
|108
|jacket
|water resistent black wind breaker
|   0.1
|109
|spare tire
|24 inch spare tire
|  22.2
|===

> the records with id: `104,105,106` will go on the same partition as the name `hammer` is the same.

Reviewer: ```suggestion
Based on the configuration, the SMT routes change events for the records that have the field name `hammer` to the same partition.
That is,  the items with `id` values `104`, `105`, and `106` are routed to the same partition.
```

// Type: reference

> // ModuleID: options-for-configuring-partition-routing-transformation
> // Title: Options for configuring partition routing transformation

Reviewer: ```suggestion
// ModuleID: options-for-configuring-the-partition-routing-transformation
// Title: Options for configuring the partition routing transformation
```

[[partition-routing-configuration-options]]
== Configuration options

The following table lists the configuration options that you can use with the partition routing SMT.

> .filter SMT configuration options

Reviewer: ```suggestion
.Partition routing SMT (`ComputePartition`) configuration options
```

[cols="30%a,25%a,45%a"]
|===
|Property
|Default
|Description

|[[partition-routing-data-collections-field-mappings]]<<partition-routing-data-collections-field-mappings, `partition.data-collections.field.mappings`>>
|

> |A comma-separated list of colon-delimited pairs to specify column used for a specific data-collection, e.g. inventory.products:name,inventory.orders:purchaser.

Reviewer: ```suggestion
|A comma-separated list of colon-delimited pairs that specify the columns to use for a specific data-collection, for example, `inventory.products:name,inventory.orders:purchaser`.
```

|[[partition-routing-data-collections-partition-num-mappings]]<<partition-routing-data-collections-partition-num-mappings, `partition.data-collections.partition.num.mappings`>>
|

> |A comma-separated list of colon-delimited pairs to specify the number of partitions used for a specific data-collection, e.g. inventory.products:2,inventory.orders:3.

Reviewer: ```suggestion
|A comma-separated list of colon-delimited pairs that specify the number of partitions to use for a specific data-collection, for example, `inventory.products:2,inventory.orders:3`.
```

:page-aliases: configuration/partition-routing.adoc
// Category: debezium-using
// Type: assembly
// ModuleID: route-record-to-specific-partition-base-on-specific-table-column
// Title: Partition Routing
[id="partition-routing"]
= Partition Routing

:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highlighter: highlight.js

toc::[]

> ifdef::community[]

Reviewer 2: This note should not be present.

// Type: concept
// Title: Example: {prodname} basic partition routing SMT configuration
// ModuleID: example-basic-debezium-partition-routing-smt-configuration
[[example-basic-partition-routing-configuration-example]]
== Example: Basic configuration

You configure the partition routing transformation in the {prodname} connector's Kafka Connect configuration.
In the configuration, you specify the tables that you are interested in, the table column you want to use to calculate the partition and the max number of partition configured for the tables.
Only events that come from configured tables will be processed. The others will be simply leaved as is.

To configure a {prodname} connector to route events to a specific partition, configure the `ComputePartition` SMT in the Kafka Connect configuration for the {prodname} connector.

For example, you might add the following configuration in your connector configuration.

[source]
----
...
topic.creation.default.partitions=2
topic.creation.default.replication.factor=1
...
transforms=ComputePartition
transforms.ComputePartition.type=io.debezium.connector.mysql.transforms.partitions.ComputePartition

> transforms.ComputePartition.partition.table.list=products,orders

Reviewer 2: Use qualified table names in examples.

Author: So you would like to have  `inventory.products?` but why only on example? I can for sure put it in the example but then I need to correctly extract the database and table name from this to correctly match the `db` and `name` filed in the `source`

Reviewer 2: @mfvitale Yes, you are right, please see other comments. It is necessary to take both db/schema name into consideration otherwise you could not handle same  table names in different schemas.

Given this `Products` table

.Products table
[cols="25%a,25%a,25%a,25%a"]
|===
|id
|name
|description
|weight

|101
|scooter
|Small 2-wheel scooter
|   3.14

|102
|car battery
|12V car battery
|   8.1
|103
|12-pack drill bits
|12-pack of drill bits with sizes ranging from #40 to #3
|   0.8
|104
|hammer
|12oz carpenter's hammer
|  0.75
|105
|hammer
|14oz carpenter's hammer
| 0.875
|106
|hammer
|16oz carpenter's hammer
|   1.0
|107
|rocks
|box of assorted rocks
|   5.3
|108
|jacket
|water resistent black wind breaker
|   0.1
|109
|spare tire
|24 inch spare tire
|  22.2
|===

the records with id: `104,105,106` will go on the same partition as the name `hammer` is the same.

// Type: reference
// ModuleID: options-for-configuring-partition-routing-transformation
// Title: Options for configuring partition routing transformation
[[partition-routing-configuration-options]]
== Configuration options

The following table lists the configuration options that you can use with the partition routing SMT.

.filter SMT configuration options
[cols="30%a,25%a,45%a"]
|===
|Property
|Default
|Description

> |[[partition-routing-topic-regex]]<<partition-routing-topic-regex, `partition.table.list`>>

Reviewer 2: The anchor names are incorrect.
